chain/indexer/src/streamer/streamer.rs

```diff
@@ -0,0 +1,231 @@
+            .map_err(|_| FailedToFetchData)
+        {
+            Ok(Ok(outcome)) => match &transaction_or_receipt_id {
+                types::TransactionOrReceiptId::Transaction { .. } => {
+                    outcomes.push(Outcome::Transaction(outcome.outcome_proof));
+                }
+                types::TransactionOrReceiptId::Receipt { .. } => {
+                    outcomes.push(Outcome::Receipt(outcome.outcome_proof));
+                }
+            },
+            _ => {
+                outcomes_to_retry.push(transaction_or_receipt_id);
+            }
+        }
+    }
+    (outcomes, outcomes_to_retry)
+}
+
+/// Fetches all the chunks by their hashes.
+/// Includes transactions and receipts in custom struct (to provide more info).
+/// Returns Chunks as a `Vec`
+async fn fetch_chunks(
+    client: &Addr<near_client::ViewClientActor>,
+    chunks: &[views::ChunkHeaderView],
+) -> Result<Vec<views::ChunkView>, FailedToFetchData> {
+    let chunks_hashes =
+        chunks.iter().map(|chunk| near_client::GetChunk::ChunkHash(chunk.chunk_hash.into()));
+    let mut chunks: futures::stream::FuturesUnordered<_> =
+        chunks_hashes.map(|get_chunk| fetch_single_chunk(&client, get_chunk)).collect();
+    let mut response = Vec::<views::ChunkView>::with_capacity(chunks.len());
+
+    while let Some(chunk) = chunks.next().await {
+        response.push(chunk?);
+    }
+
+    Ok(response)
+}
+
+/// Function that starts Streamer's busy loop. Every half a seconds it fetches the status
+/// compares to already fetched block height and in case it differs fetches new block of given height.
+///
+/// We have to pass `client: Addr<near_client::ClientActor>` and `view_client: Addr<near_client::ViewClientActor>`.
+pub(crate) async fn start(
+    view_client: Addr<near_client::ViewClientActor>,
+    client: Addr<near_client::ClientActor>,
+    mut blocks_sink: mpsc::Sender<BlockResponse>,
+) {
+    info!(target: INDEXER, "Starting Streamer...");
+    let mut outcomes_to_get = Vec::<types::TransactionOrReceiptId>::new();
+    let mut last_synced_block_height: types::BlockHeight = 0;
+    'main: loop {
+        time::delay_for(INTERVAL).await;
+        let status = fetch_status(&client).await;
+        if let Ok(status) = status {
+            if status.sync_info.syncing {
+                continue;
+            }
+        }
+
+        let block = if let Ok(block) = fetch_latest_block(&view_client).await {
+            block
+        } else {
+            continue;
+        };
+
+        let latest_block_height = block.header.height;
+        if last_synced_block_height == 0 {
+            last_synced_block_height = latest_block_height;
+        }
+        debug!(
+            target: INDEXER,
+            "The last synced block is #{} and the latest block is #{}",
+            last_synced_block_height,
+            latest_block_height
+        );
+        for block_height in (last_synced_block_height + 1)..=latest_block_height {
```

Comment:
it seems that something will break if we don't sync all blocks within the range. For example you will have outgoing receipts whose execution outcome cannot be found.

Reply:
What would be the desired behavior in such a case? I would prefer to have sync working and stall the indexing until we sync all the necessary data for the block. Is it possible?

Reply:
It is possible but that requires some special configurations. Also it will be extremely slow to sync if we do that. I guess my question is whether some internal logic of indexer will break if certain data are not available.

Reply:
Well, it will be suboptimal if indexer does not live up to the expectations that it streams every single block (currently, we designed it to at least work for the blocks that we observe on the fly, but it would be great to be able to serve those reliably, so the users don’t need to workaround the possibility of missing (those are produced by the network, but missed by the indexer node)). What configuration is that you are talking about?

Reply:
You can set `block_fetch_horizon` to `u64::MAX` to force block sync. However, as I said that would be excruciatingly slow. I am not sure whether we really want to do that. Also I realize an issue: how do we update account state? Do we only update it if we see that a transaction/receipt touches it? If so that is flawed and could be potentially misleading.

Reply:
@bowenwang1996 it is really bad that we use 64-bit integers in JSON, meh...

Reply:
@bowenwang1996 do you suggest we also include the state changes? I wonder how we should display the rewards (balance increase due to validation duty) and kickouts (stake gets automatically returned to the balance) on Explorer :thinking: Created a separate issue to discuss this: https://github.com/near/near-explorer/issues/354

Reply:
*Just for the future reference, the further work related to `block_fetch_horizon` is discussed in #3022.*